### PR TITLE
feat(rpc): use log indices for address and topic filters

### PR DESCRIPTION
**Pull request by rkrasiuk**
Builds on top of #3082.

## Description

This PR modifies the matching logic of `address` and `topic` filters in `eth_getLogs` handler.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/filter.rs [hunk 1]
```diff
@@ -377,48 +382,45 @@ where
...
+                    Some(header) => header,
+                    None => continue,
+                };
+
                 // these are consecutive headers, so we can use the parent hash of the next block to
                 // get the current header's hash
                 let num_hash: BlockHashOrNumber = headers
                     .get(idx + 1)
-                    .map(|h| h.parent_hash.into())
+                    .map(|h| h.as_ref().unwrap().parent_hash.into())
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/filter.rs:406]: Let's add a note on safety of the unwrap due to the continue above

File: crates/storage/provider/src/providers/database/provider.rs [hunk 2]
```diff
@@ -350,6 +350,45 @@ impl<'this, TX: DbTx<'this>> DatabaseProvider<'this, TX> {
             .map(|address| plain_accounts.seek_exact(address).map(|a| (address, a.map(|(_, v)| v))))
             .collect::<std::result::Result<Vec<_>, _>>()?)
     }
+
+    /// Get history indices in block range
+    pub fn history_indices_in_block_range<K, T>(
```
Comment (gakonst) [inline crates/storage/provider/src/providers/database/provider.rs:355]: Can we test this?

File: crates/rpc/rpc/src/eth/filter.rs [hunk 3]
```diff
@@ -428,6 +430,116 @@ where
...
+                self.union_indices(address_indices);
+            }
+        }
+        Ok(())
+    }
+
+    fn install_topic_filter(
+        &mut self,
+        provider: &impl LogIndexProvider,
+        topic_filters: Option<&Vec<ValueOrArray<Option<H256>>>>,
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/filter.rs:518]: Cleaner api if it didn't take an option given it doesn't do anything if None

File: crates/rpc/rpc/src/eth/filter.rs [hunk 4]
```diff
@@ -428,6 +430,116 @@ where
...
+        }
+    }
+
+    /// Returns `true` if the filter is active but has no matching block numbers.
+    fn is_empty(&self) -> bool {
+        matches!(self.indices, Some(None))
+    }
+
+    /// Change the inner indices to be the union of itself and incoming.
+    fn union_indices(&mut self, new_indices: Option<IntegerList>) {
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/filter.rs:471]: Intersection?

**Review by gakonst — CHANGES_REQUESTED** [review]
Very nice!

File: crates/primitives/src/integer_list.rs [hunk 5]
```diff
@@ -69,6 +69,37 @@ impl IntegerList {
     pub fn from_bytes(data: &[u8]) -> Result<Self, EliasFanoError> {
         Ok(Self(EliasFano::deserialize_from(data).map_err(|_| EliasFanoError::FailedDeserialize)?))
     }
+
+    /// Iterates over two lists of integers and creates a union.
+    ///
+    /// Returns `None` if the resulting list is empty.
+    pub fn union(&self, other: &Self) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/primitives/src/integer_list.rs:76]: ah, lol, you're right, that's an intersection

File: crates/rpc/rpc/src/eth/filter.rs (hunk 1 above)
Comment (rkrasiuk) [reply crates/rpc/rpc/src/eth/filter.rs:406]: looked over this again, it's not actually safe and is sometimes incorrect. fixing

File: crates/rpc/rpc/src/eth/filter.rs [hunk 6]
```diff
@@ -377,48 +382,45 @@ where
 
         let topics = filter.has_topics().then(|| filter_params.flat_topics.clone());
 
-        // derive bloom filters from filter input
-        let address_filter = FilteredParams::address_filter(&filter.address);
-        let topics_filter = FilteredParams::topics_filter(&topics);
+        // Create log index filter
+        let mut log_index_filter =
+            LogIndexFilter::new(from_block..=to_block, self.max_headers_range);
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/rpc/rpc/src/eth/filter.rs:387]: @mattsse what's a reasonable default value for `max_headers_range`?

**Review by onbjerg — APPROVED** [review]
lgtm modulo missing sane default mentioned above, ping @mattsse 

might be worth just adding a default of something <= the headers download stage given we know how much memory that uses and then tune based on performance readings?

**Review by mattsse — CHANGES_REQUESTED** [review]
a few nits, the rpc changes look good.

File: crates/rpc/rpc/src/eth/filter.rs [hunk 7]
```diff
@@ -377,48 +382,45 @@ where
...
+        log_index_filter.install_address_filter(&self.provider, filter.address.as_ref())?;
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/rpc/rpc/src/eth/filter.rs]: install is better since now these calls are preceded by `if let Some` check

**Pull request merged**